packages/core/src/pubsub/pubsub.ts

```diff
@@ -0,0 +1,86 @@
+import { Observable, EMPTY, pipe, of, from, Subscription, UnaryFunction } from 'rxjs';
+import { deserialize, PubsubMessage, serialize } from './pubsub-message';
+import { IpfsApi } from '@ceramicnetwork/common';
+import { map, catchError, mergeMap, withLatestFrom } from 'rxjs/operators';
+import { IncomingChannel, filterOuter, IPFSPubsubMessage } from './incoming-channel';
+import { DiagnosticsLogger, ServiceLogger } from '@ceramicnetwork/logger';
+import { Memoize } from 'typescript-memoize';
+
+/**
+ * Deserialize incoming message in an internal observable that does not emit if error happens.
+ * Log a successfully deserialized message.
+ *
+ * @param peerId$ - own IPFS node peer id
+ * @param pubsubLogger - logger that dumps a successfully deserialized message
+ * @param topic - IPFS pubsub topic we listen
+ */
+function ipfsToPubsub(
+  peerId$: Observable<string>,
+  pubsubLogger: ServiceLogger,
+  topic: string,
+): UnaryFunction<Observable<IPFSPubsubMessage>, Observable<PubsubMessage>> {
+  return pipe(
+    withLatestFrom(peerId$),
+    mergeMap(([incoming, peerId]) =>
+      of(incoming).pipe(
+        map((incoming) => {
+          const message = deserialize(incoming);
+          const logMessage = { ...incoming, data: message };
+          delete logMessage.key;
+          delete logMessage.signature;
+          pubsubLogger.log({ peer: peerId, event: 'received', topic: topic, message: logMessage });
+          return message;
+        }),
+        catchError(() => EMPTY),
+      ),
+    ),
+  );
+}
+
+/**
+ * Receive and publish messages to IPFS pubsub.
+ */
+export class Pubsub extends Observable<PubsubMessage> {
+  constructor(
+    private readonly ipfs: IpfsApi,
+    private readonly topic: string,
+    private readonly resubscribeEvery: number,
+    private readonly pubsubLogger: ServiceLogger,
+    private readonly logger: DiagnosticsLogger,
+  ) {
+    super((subscriber) => {
+      const incoming$ = new IncomingChannel(ipfs, topic, resubscribeEvery, pubsubLogger, logger);
+
+      incoming$
+        .pipe(filterOuter(this.peerId$), ipfsToPubsub(this.peerId$, pubsubLogger, topic))
+        .subscribe(subscriber);
+    });
+  }
+
+  @Memoize()
+  private get peerId$() {
+    return from<Promise<string>>(this.ipfs.id().then((_) => _.id));
+  }
+
+  /**
+   * Publish message to IPFS Pubsub as a "fire and forget" command.
+   * Returned Subscription is free to ignore.
```

Comment:
is there ever a case in which you would use it?  If so, how?  If not, why return it at all?

Reply:
Maybe you would like to do something exactly after the message is published.

Reply:
Like in https://github.com/ceramicnetwork/js-ceramic/pull/1058/files/2729068e8f050414983bb508add23123680246b4#diff-0e6de57da2fc7ba36929ce8a926162b43593faa16fe1750c83a5e5c622560dfcR106

Reply:
Makes sense, but can we just add a message to the comment clarifying that?

Reply:
Yup.